### PR TITLE
Run Rust tests on MacOS in CI

### PR DESCRIPTION
**Pull request by thomaseizinger**
Fixes https://github.com/comit-network/comit-rs/issues/2927.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
> so now we need to run `cargo test --features testcontainers` locally to ensure that all test pass. Can't even run `make test` for that. Can we have at least a makefile shortcut?

I didn't adapt the Makefile because you argued in the past that the Makefile should do the same thing on CI and locally.

We are generally erring on the side of using the Makefile to optimize the local development environment. I am happy to change the Makefile but that implies for me that we shouldn't use it in CI anymore if we want to be consistent.

I would prefer that to be honest. Primarily because CI and local have very different needs in terms of optimiztion and compared to languages like C or C++, cargo is a powerful enough package manager to act as a solid base layer that makes sure everything is compiled accordingly.

That is, as long as we stick to cargo's conventions as much as possible. We do that already, hence our Makefile is just a thin wrapper around cargo anyway.

> I prefer the way I did with nectar where one can `deactivate` the docker tests if they can't run docker. I understand this is not possible to do the feature way (as you mentioned that dev-dependencies cannot be optional).

Using features this way can be tricky though. While it may not be relevant in this specific case, deactivating default features only works if every crate in the dependency chain opts out of the default features.
Hence, I'd prefer if we stick to cargo's conventions here and make sure that building and testing works on every platform _by default_ and users can opt-in to more things.

An unsolved problem of not using the Makefile in CI is how we want to deal with applying and checking the formatting without too much duplication. I don't have an answer to this.

Generally, I'd happy if we can merge this before solving all these problems.
The presented patch makes our CI more robust at the expense of a minor annoyance if you want to run that **one** additional testcontainers test.

File: comit/Cargo.toml [hunk 1]
```diff
@@ -34,6 +34,9 @@ serdebug = "1"
 sha2 = "0.9"
 strum = "0.18.0"
 strum_macros = "0.18.0"
+# dev-dependencies can't be optional :(
```
Comment (bonomat) [inline comit/Cargo.toml:37]: `can't be optional` but `optional=true`? I don't understand :)

**Comment by D4nte** [comment]
> Hence, I'd prefer if we stick to cargo's conventions here and make sure that building and testing works on every platform by default and users can opt-in to more things.

I'd be curious to have a source to this statement. If it were the case then `std` would be opt-in and not opt-out in most crates.

**Comment by thomaseizinger** [comment]
> > Hence, I'd prefer if we stick to cargo's conventions here and make sure that building and testing works on every platform by default and users can opt-in to more things.
> 
> I'd be curious to have a source to this statement. If it were the case then `std` would be opt-in and not opt-out in most crates.

I can't find a specific source for it but it is basically the result of logical deduction around how features are designed in cargo.

Features are accumulated across the dependency graph, which means as soon as one crate opts in to a feature, it is impossible to remove it from the graph.

Most crates have a "no-std" feature that actually _opts-in_ to not having `std`.

I can't find the thread at the moment but I've read somewhere that the Rust team actually intends to modularise `std` with cargo features so there is a more fine-grained choice in what you get. The reasoning behind that is `no_std` removes too much and it doesn't clearly map to certain environments. It works for micro-controllers but then there is also cases where you actually can allocate but you don't have other things that you get with `std`.

Once that is achieved, crates opt-in to only what they need and expose more features to add more functionality instead of adding a feature that removes all of `std`.

File: comit/Cargo.toml (hunk 1 above)
Comment (thomaseizinger) [reply comit/Cargo.toml:37]: Read the comment again :)

"dev-dependencies can't be optional" - hence this dependency is under `[dependencies]` although it is just used for tests ;)

Comment (bonomat) [reply comit/Cargo.toml:37]: aahhh... got it :D

**Pull request closed without merging**